### PR TITLE
Fix top-level window size for non-linear dimensions based on AutoScaleMode

### PR DESCRIPTION
**Pull request by dreddy-work**
In PermonV2 mode applications, We have been using linear sizes for top-level windows that are provided by the Windows irrespective of the AutoScaleMode.  This has been problematic for AutoScaleMode.Font, where Form scaling should be non-linear given its child controls are scaled non-linearly and depend on the Font that was assigned to Form/child controls.

We are making changes to utilize [GetDpiScaledSize](https://learn.microsoft.com/en-us/windows/win32/hidpi/wm-getdpiscaledsize) message and let Windows know that Form may need non-linear sizes depending on AutoScaleMode.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7973)

### CODE REVIEW

**Comment by dreddy-work** [comment]
Tests pass locally. Need to investigate.

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs [hunk 1]
```diff
@@ -143,12 +143,15 @@ protected SizeF AutoScaleFactor
...
                 // is the unit scale.
-                _currentAutoScaleFactor = saved.IsEmpty ? new SizeF(1.0F, 1.0F) : new SizeF(current.Width / saved.Width, current.Height / saved.Height);
+                _currentAutoScaleFactor = GetCurrentAutoScaleFactor(current, saved);
 
                 return _currentAutoScaleFactor;
             }
         }
 
+        internal static SizeF GetCurrentAutoScaleFactor(SizeF currentAutoScaleDimensions, SizeF savedAutoScaleDimensions)
+            => savedAutoScaleDimensions.IsEmpty ? new SizeF(1.0F, 1.0F) : new SizeF(currentAutoScaleDimensions.Width / savedAutoScaleDimensions.Width, currentAutoScaleDimensions.Height / savedAutoScaleDimensions.Height);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:153]: [nit] this looks like a very long line (even by my standards, which are lax). Do you have [Editor guidelines](https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines) extension installed?

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs [hunk 2]
```diff
@@ -1461,16 +1470,8 @@ internal void ScaleContainerForDpi(int deviceDpiNew, int deviceDpiOld, Rectangle
...
-
-                // DpiFontscache is available only in PermonitorV2 mode applications.
-                if (!TryGetDpiFont(deviceDpiNew, out Font? fontForDpi))
-                {
-                    Font currentFont = Font;
-                    fontForDpi = currentFont.WithSize(currentFont.Size * factor);
-                    AddToDpiFonts(deviceDpiNew, fontForDpi);
-                }
+                TryGetDpiFont(deviceDpiNew, out Font? fontForDpi);
+                Debug.Assert(fontForDpi != null, "We should have Font updated for Dpi from WM_GETDPISCALEDSIZE message");
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1474]:
```suggestion
                Debug.Assert(fontForDpi is not null, "Font should have been updated for DPI from WM_GETDPISCALEDSIZE message");
```

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1474]: `TryGetDpiFont(..)` indicates whether it's successful or not. Why aren't we using the result and throwing, if it's unexpected?

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 3]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
         [EditorBrowsable(EditorBrowsableState.Advanced)]
         protected virtual bool OnGetDpiScaledSize(int deviceDpiOld, int deviceDpiNew, ref Size desiredSize)
         {
-            return false; // scale linearly
+            // Compute and update Font for the current Dpi.
+            var factor = ((float)deviceDpiNew) / deviceDpiOld;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4325]: Please use the type here

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 4]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
         [EditorBrowsable(EditorBrowsableState.Advanced)]
         protected virtual bool OnGetDpiScaledSize(int deviceDpiOld, int deviceDpiNew, ref Size desiredSize)
         {
-            return false; // scale linearly
+            // Compute and update Font for the current Dpi.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4324]: This comment is either misplaces or inaccurate. The adjacent code calculates the scaling factor.

Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4324]: Here and other comments - the casing appears rather arbitrary. DPI is an acronym and should be capitalized. The "font" doesn't appear to carry any special meaning that could warrant capitalization.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 5]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
...
+
+            // Dpi specific fonts cache is available only in PermonitorV2 mode applications.
+            if (!TryGetDpiFont(deviceDpiNew, out Font? fontForDpi))
+            {
+                Font currentFont = Font;
+                fontForDpi = currentFont.WithSize(currentFont.Size * factor);
+                AddToDpiFonts(deviceDpiNew, fontForDpi);
+            }
+
+            // If AutoScaleMode is Dpi, We continue with the linear size we get from Windows for the top-level window.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4335]:
```suggestion
            // If we DPI-scaling then we continue with the linear size we get from Windows for the top-level window.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 6]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
...
+            // need non-linear size for top-level window based on AutoScaleMode property.
+            FontHandleWrapper fontwrapper = new FontHandleWrapper(fontForDpi);
+            SizeF currentAutoScaleDimensions = GetCurrentAutoScaleDimensions(fontwrapper.Handle);
+            SizeF autoScaleFactor = GetCurrentAutoScaleFactor(currentAutoScaleDimensions, AutoScaleDimensions);
+
+            desiredSize.Width = (int)(Size.Width * autoScaleFactor.Width);
+            desiredSize.Height = (int)(Size.Height * autoScaleFactor.Height);
+            Debug.WriteLine($"AutoScaleFactor computed for new Dpi = {autoScaleFactor.Width} - {autoScaleFactor.Height}");
+
+            return true; // Notifying Windows that the top-level window size should be based on AutoScale mode.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4352]:
```suggestion
            // Notify Windows that the top-level window size should be based on AutoScaleMode value.
            return true; 
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs (hunk 2 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1474]: We won't see it null in any practical/real-time scenarios but when writing tests, if you don't send WM_GETDPISCALEDSIZE message but send WM_DPICHNAGED directly, you will end up in this state. Debug assertion is to help those test scenarios.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 3 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4325]: i don't insist but as per our guidelines, type is clear here to use `var`.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 5 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4335]: I still feel my sentence is better describing it.  `we DPI-scaling` is ambiguous and doesn't tell how that happen. Any better way to include `AutoScaleMode` property in the description to make it clear on what type of scaling is coming form?

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1474]: Right, makes sense. Perhaps it's worth adding a comment here.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 5 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4335]: Makes sense. Perhaps then `If AutoScaleMode=AutoScaleMode.Dpi...`

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 7]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
         [EditorBrowsable(EditorBrowsableState.Advanced)]
         protected virtual bool OnGetDpiScaledSize(int deviceDpiOld, int deviceDpiNew, ref Size desiredSize)
         {
-            return false; // scale linearly
+            float factor = ((float)deviceDpiNew) / deviceDpiOld;
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4324]: I am wondering:
When a Form is a) used as a control or b) is an MDI-Form, does that also work the intended way?

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 8]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
...
+
+            // Compute font for the current DPI and cache it. DPI specific fonts cache is available only in PermonitorV2 mode applications.
+            if (!TryGetDpiFont(deviceDpiNew, out Font? fontForDpi))
+            {
+                Font currentFont = Font;
+                fontForDpi = currentFont.WithSize(currentFont.Size * factor);
+                AddToDpiFonts(deviceDpiNew, fontForDpi);
+            }
+
+            // If AutoScaleMode=AutoScaleMode.Dpi, We continue with the linear size we get from Windows for the top-level window.
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4334]:
```suggestion
            // If AutoScaleMode=AutoScaleMode.Dpi then we continue with the linear size we get from Windows for the top-level window.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs [hunk 9]
```diff
@@ -1461,16 +1472,10 @@ internal void ScaleContainerForDpi(int deviceDpiNew, int deviceDpiOld, Rectangle
...
+                TryGetDpiFont(deviceDpiNew, out Font? fontForDpi);
 
-                // DpiFontscache is available only in PermonitorV2 mode applications.
-                if (!TryGetDpiFont(deviceDpiNew, out Font? fontForDpi))
-                {
-                    Font currentFont = Font;
-                    fontForDpi = currentFont.WithSize(currentFont.Size * factor);
-                    AddToDpiFonts(deviceDpiNew, fontForDpi);
-                }
+                // For test scenarios, if we send WM_DPICHNAGED message but not WM_GETDPISCALEDSIZE message, fontForDpi may be null.
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1477]:
```suggestion
                // For test scenarios, if we send WM_DPICHANGED message but not WM_GETDPISCALEDSIZE message, fontForDpi may be null.
```

**Comment by KlausLoeffelmann** [comment]
> We have been using linear sizes for top-level windows that are provided by the Windows irrespective of the AutoScaleMode.

I don't understand what that means. Do you have a screenshot or a GIF which could illustrate that?

**Review by KlausLoeffelmann — COMMENTED** [review]
Have a question and a few typo nits.

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs (hunk 7 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4324]: In those cases, it is not top-level window and does not receive these messages. They only get DPI_CHANGED_BEEFORE and AFTER events.

**Comment by dreddy-work** [comment]
> > We have been using linear sizes for top-level windows that are provided by the Windows irrespective of the AutoScaleMode.
> 
> I don't understand what that means. Do you have a screenshot or a GIF which could illustrate that?

Gif may not justify but let me try describing this:

When you move Form from 100% monitor to 200% monitor,  factor become 2 for both `width` and `height`. Windows use this factor and scale the Form. However, when AutoScaleMode is `Font`, for the same move operation, Factors will not be linear and have separate factors for `Width` and `Height` (i.e  width factor - 1.1 and height factor 0.95 etc) .

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs [hunk 10]
```diff
@@ -143,12 +143,17 @@ protected SizeF AutoScaleFactor
...
+                _currentAutoScaleFactor = GetCurrentAutoScaleFactor(current, saved);
 
                 return _currentAutoScaleFactor;
             }
         }
 
+        internal static SizeF GetCurrentAutoScaleFactor(SizeF currentAutoScaleDimensions, SizeF savedAutoScaleDimensions)
+            => savedAutoScaleDimensions.IsEmpty ?
+            new SizeF(1.0F, 1.0F)
+            : new SizeF(currentAutoScaleDimensions.Width / savedAutoScaleDimensions.Width, currentAutoScaleDimensions.Height / savedAutoScaleDimensions.Height);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:155]: [nit]
```suggestion
            => savedAutoScaleDimensions.IsEmpty
                ? new(1.0F, 1.0F)
                : new(currentAutoScaleDimensions.Width / savedAutoScaleDimensions.Width, currentAutoScaleDimensions.Height / savedAutoScaleDimensions.Height);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs [hunk 11]
```diff
@@ -1461,16 +1472,10 @@ internal void ScaleContainerForDpi(int deviceDpiNew, int deviceDpiOld, Rectangle
...
 
-                // DpiFontscache is available only in PermonitorV2 mode applications.
-                if (!TryGetDpiFont(deviceDpiNew, out Font? fontForDpi))
-                {
-                    Font currentFont = Font;
-                    fontForDpi = currentFont.WithSize(currentFont.Size * factor);
-                    AddToDpiFonts(deviceDpiNew, fontForDpi);
-                }
+                // For test scenarios, if we send WM_DPICHNAGED message but not WM_GETDPISCALEDSIZE message, fontForDpi may be null.
+                Debug.Assert(fontForDpi != null, "We should have Font updated for Dpi from WM_GETDPISCALEDSIZE message");
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ContainerControl.cs:1478]:
```suggestion
                Debug.Assert(fontForDpi is not null, "Font should have been updated for DPI from WM_GETDPISCALEDSIZE message");```

File: src/System.Windows.Forms/src/System/Windows/Forms/Form.cs [hunk 12]
```diff
@@ -4321,7 +4321,35 @@ private void WmDpiChanged(ref Message m)
...
+            // Calculate AutoscaleFactor for AutoScaleMode.Font. We will be using this factor to scale child controls
+            // and use same factor to compute desired size for top-level windows for the current DPI.
+            // This desired size is then used to notify Windows that we need non-linear size for top-level window.
+            FontHandleWrapper fontwrapper = new FontHandleWrapper(fontForDpi);
+            SizeF currentAutoScaleDimensions = GetCurrentAutoScaleDimensions(fontwrapper.Handle);
+            SizeF autoScaleFactor = GetCurrentAutoScaleFactor(currentAutoScaleDimensions, AutoScaleDimensions);
+
+            desiredSize.Width = (int)(Size.Width * autoScaleFactor.Width);
+            desiredSize.Height = (int)(Size.Height * autoScaleFactor.Height);
+            Debug.WriteLine($"AutoScaleFactor computed for new Dpi = {autoScaleFactor.Width} - {autoScaleFactor.Height}");
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Form.cs:4349]:
```suggestion
            Debug.WriteLine($"AutoScaleFactor computed for new DPI = {autoScaleFactor.Width} - {autoScaleFactor.Height}");
```

**Review by RussKie — APPROVED** [review]
👍 with few nits

**Review by RussKie — APPROVED** [review]

**Comment by dreddy-work** [comment]
Thank you.

**Comment by dreddy-work** [comment]
https://github.com/dotnet/docs/issues/33696

**Pull request merged**